src/lib/contract_metadata.ml

```diff
@@ -0,0 +1,209 @@
+(*****************************************************************************)
+(*                                                                           *)
+(* Open Source License                                                       *)
+(* Copyright (c) 2020 TQ Tezos <contact@tqtezos.com>                         *)
+(*                                                                           *)
+(* Permission is hereby granted, free of charge, to any person obtaining a   *)
+(* copy of this software and associated documentation files (the "Software"),*)
+(* to deal in the Software without restriction, including without limitation *)
+(* the rights to use, copy, modify, merge, publish, distribute, sublicense,  *)
+(* and/or sell copies of the Software, and to permit persons to whom the     *)
+(* Software is furnished to do so, subject to the following conditions:      *)
+(*                                                                           *)
+(* The above copyright notice and this permission notice shall be included   *)
+(* in all copies or substantial portions of the Software.                    *)
+(*                                                                           *)
+(* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR*)
+(* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,  *)
+(* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL   *)
+(* THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER*)
+(* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING   *)
+(* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER       *)
+(* DEALINGS IN THE SOFTWARE.                                                 *)
+(*                                                                           *)
+(*****************************************************************************)
+
+open! Base
+open Import
+
+module Uri = struct
+  let validate uri_code =
+    let open Metadata_uri in
+    let errors = ref [] in
+    let error w src e = errors := (w, src, e) :: !errors in
+    let validate_kt1_address s =
+      ( try ignore (B58_hashes.check_b58_kt1_hash s) with
+      | Failure f -> error `Address s f
+      | e -> Fmt.kstr (error `Address s) "%a" Exn.pp e ) ;
+      Ok () in
+    let validate_network = function
+      | "mainnet" | "carthagenet" | "delphinet" | "dalphanet" | "zeronet" ->
+          Ok ()
+      | s ->
+          ( try ignore (B58_hashes.check_b58_chain_id_hash s) with
+          | Failure f -> error `Network s f
+          | e -> Fmt.kstr (error `Network s) "%a" Exn.pp e ) ;
+          Ok () in
+    let uri =
+      Uri.of_string uri_code |> of_uri ~validate_kt1_address ~validate_network
+    in
+    (uri, List.rev !errors)
+
+  let%test "validate_address bogus-schema" =
+    (* Here, we expect  validate_address to return Ok even though it has some errors.  FIXME: why?*)
```

Comment:
Why does this test pass?  I don't mean from a code perspective -- I mean why does a function named validate_address return success here?  Should we fix this?

Reply:
It does indeed deserve a better name and/or a better structured return value. But in that case it should return `Error _, _`; a wrong scheme is a hard failure.

See
https://github.com/oxheadalpha/tezos-contract-metadata/blob/4c1f5dc0b2d27d1c096c78bf49ea36e0975f6492/src/lib/metadata_uri.ml#L196